lib/logstash/runner.rb

```diff
@@ -30,95 +85,245 @@ def main(args)
     # Print a warning to STDERR for bad java versions
     LogStash::Util::JavaVersion.warn_on_bad_java_version
 
-    Stud::untrap("INT", @startup_interruption_trap)
+    if version?
+      show_version
+      return 0
+    end
+
+    configure
+
+    return start_shell(@ruby_shell, binding) if @ruby_shell
 
-    task = run(args)
-    exit(task.wait)
+    if @config_string.nil? && @config_path.nil?
+      fail(I18n.t("logstash.runner.missing-configuration"))
+    end
+
+    @agent.logger = @logger
+
+    config_string = format_config(@config_path, @config_string)
+
+    @agent.add_pipeline("base", config_string, :filter_workers => filter_workers)
+
+    if config_test?
+      puts "Configuration OK"
+    else
+      task = Stud::Task.new { @agent.execute }
+      return task.wait
+    end
+
+  rescue LoadError => e
+    fail("Configuration problem.")
+  rescue LogStash::ConfigurationError => e
+    @logger.fatal I18n.t("logstash.error", :error => e)
+    if !config_test?
+      @logger.warn I18n.t("logstash.runner.configtest-flag-information")
+    end
+    show_short_help
+    return 1
+  rescue => e
```

Comment:
Should this be `rescue Exception` now given @guyboertje and @colinsurprenant 's latest research?